### PR TITLE
feat(asm): add IP blocking support for Django

### PR DESCRIPTION
**Pull request by juanjux**
Note: this builds over and targets -> https://github.com/DataDog/dd-trace-py/pull/5032

# Description

Adds support for IP Blocking for the Django framework.

# Motivation

Having Django users be able to block IPs.

# Design

This builds over the design outlined in https://github.com/DataDog/dd-trace-py/pull/5032. For Django, the implementation stores the IP and header values using the new ASM context manager on the `traced_get_response` method and in the same function it checks if the request was blocked, producing an `HttpResponseForbidden` respone. The changes in the other files are for re-using the IP and headers if set by `AppsecSpanProcessor.on_span_start`.

This also supports IP blocking under ASGI. This was added because the Django Snapshot tests use ASGI.

# Testing strategy

- Have snapshot tests.
- Has gone trough two interactive ASM peer testing sessions.
- Checked manually against system tests (`tests/appsec/test_ip_blocking.py`). System tests for this feature in CI will be enabled once we have a released version with this PR merged.

Signed-off-by: Juanjo Alvarez <juanjo.alvarezmartinez@datadoghq.com>

## Checklist

- [X] Change(s) are motivated and described in the PR description.
- [X] Testing strategy is described if automated tests are not included in the PR.
- [X] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [X] Change is maintainable (easy to change, telemetry, documentation).
- [X] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [X] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).

Note: public docs will be released once the beta is publicly announced.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.

### CODE REVIEW

File: ddtrace/contrib/asgi/middleware.py [hunk 1]
```diff
@@ -111,104 +123,136 @@ async def __call__(self, scope, receive, send):
...
+        else:
+            ip = ""
 
-        span = self.tracer.trace(
-            name=self.integration_config.get("request_span_name", "asgi.request"),
-            service=trace_utils.int_service(None, self.integration_config),
-            resource=resource,
-            span_type=SpanTypes.WEB,
-        )
+        resource = "{} {}".format(scope["method"], scope["path"])
```
Comment (P403n1x87) [inline ddtrace/contrib/asgi/middleware.py:132]:
```suggestion
        resource = " ".join((scope["method"], scope["path"]))
```

**Review by P403n1x87 — APPROVED** [review]

**Review by avara1986 — APPROVED** [review]

File: tests/appsec/test_remoteconfiguration.py [hunk 2]
```diff
@@ -156,6 +156,7 @@ def test_rc_activation_ip_blocking_data(tracer, remote_config_worker):
                 set_http_meta(
                     span,
                     Config(),
+                    request_headers={},
```
Comment (emmettbutler) [inline tests/appsec/test_remoteconfiguration.py:159]: Why is this necessary?

File: tests/contrib/django/test_django_snapshots.py [hunk 3]
```diff
@@ -284,3 +289,90 @@ def test_templates_disabled():
         resp = client.get("/template-view/")
         assert resp.status_code == 200
         assert resp.content == b"some content\n"
+
+
+@pytest.mark.skipif(django.VERSION < (3, 2, 0), reason="Only want to test with latest Django")
```
Comment (emmettbutler) [inline tests/contrib/django/test_django_snapshots.py:294]: How long will it be true that this mark filters to the "latest" Django version?

File: ddtrace/internal/remoteconfig/client.py [hunk 4]
```diff
@@ -239,14 +239,14 @@ def base64_to_struct(val, cls):
         self.converter.register_structure_hook(SignedRoot, base64_to_struct)
         self.converter.register_structure_hook(SignedTargets, base64_to_struct)
 
-        self._products = dict()  # type: MutableMapping[str, ProductCallback]
+        self._products = dict()  # type: MutableMapping[str, Optional[ProductCallback]]
```
Comment (emmettbutler) [inline ddtrace/internal/remoteconfig/client.py:242]: Does this mean that keys in this dictionary can point to `None` values? It's not clear to me how the changes in this file are related to the rest of the diff.

**Review by emmettbutler — CHANGES_REQUESTED** [review]

File: ddtrace/internal/remoteconfig/client.py (hunk 4 above)
Comment (juanjux) [reply ddtrace/internal/remoteconfig/client.py:242]: None, vestigial from the previous version of the base branch, removed.

File: tests/appsec/test_remoteconfiguration.py (hunk 2 above)
Comment (juanjux) [reply tests/appsec/test_remoteconfiguration.py:159]: Vestigial. Removed.

File: tests/contrib/django/test_django_snapshots.py (hunk 3 above)
Comment (juanjux) [reply tests/contrib/django/test_django_snapshots.py:294]: Dunno. But it's in line with other tests like it. I guess they could be renamed to "latest Djangos" in other PR.

**Comment by juanjux** [comment]
After a nice 6 hours bugfixing session CI is green, PTAL.

**Review by gnufede — DISMISSED** [review]

**Comment by juanjux** [comment]
@emmettbutler could you check if the latest changes are ok for your feedback? (you have a change request but I already did changes or replied to the feedback).

**Review by emmettbutler — APPROVED** [review]

**Review by avara1986 — APPROVED** [review]

**Review by christophe-papazian — DISMISSED** [review]

**Review by emmettbutler — DISMISSED** [review]

**Review by avara1986 — APPROVED** [review]

**Review by emmettbutler — APPROVED** [review]
